plugins/cassandra/cassandrawu.cpp

```diff
@@ -2470,6 +2304,169 @@ class CCassandraWorkUnit : public CPersistedWorkUnit
+        const char *wuid = queryWuid();
+        statement.bindInt32(0, rtlHash32VStr(wuid, 0) % NUM_PARTITIONS);
+        statement.bindString(1, wuid);
+        statement.bindString(2, graphName);
+        CassandraFuture future(cass_session_execute(sessionCache->querySession(), statement));
+        future.wait("select state");
+        CassandraResult result(cass_future_get_result(future));
+        CassandraIterator rows(cass_iterator_from_result(result));
+        if (!cass_result_row_count(result))
+            return NULL;
+        Owned<IPropertyTree> progress = createPTree(graphName);
+        progress->setPropBool("@stats", true);
+        progress->setPropInt("@format", PROGRESS_FORMAT_V);
+        while (cass_iterator_next(rows))
+        {
+            const CassRow *row = cass_iterator_get_row(rows);
+            unsigned subId = subId = getUnsignedResult(NULL, cass_row_get_column(row, 0));
+            StringBuffer creator, xml;
+            getCassString(creator, cass_row_get_column(row, 1));
+            getCassString(xml, cass_row_get_column(row, 2));
+            IPTree *stats = createPTreeFromXMLString(xml);
+            // We could check that atoi(stats->queryName()+2)==subgraphID, and that stats->queryProp(@creator)==creator)....
+            progress->addPropTree(stats->queryName(), stats);
+        }
+        return createConstGraphProgress(queryWuid(), graphName, progress); // Links progress
+    }
+    WUGraphState queryGraphState(const char *graphName) const
+    {
+        return queryNodeState(graphName, 0);
+    }
+    WUGraphState queryNodeState(const char *graphName, WUGraphIDType nodeId) const
+    {
+        CassandraStatement statement(sessionCache->prepareStatement("SELECT state FROM wuGraphState where partition=? and wuid=? and graphID=? and subgraphID=?;"));
+        const char *wuid = queryWuid();
+        statement.bindInt32(0, rtlHash32VStr(wuid, 0) % NUM_PARTITIONS);
+        statement.bindString(1, wuid);
+        statement.bindString(2, graphName);
+        statement.bindInt32(3, nodeId);
+        CassandraFuture future(cass_session_execute(sessionCache->querySession(), statement));
+        future.wait("select state");
+        CassandraResult result(cass_future_get_result(future));
+        if (cass_result_row_count(result))
+            return (WUGraphState) getUnsignedResult(NULL, getSingleResult(result));
+        else
+            return WUGraphUnknown;
+    }
+    void setGraphState(const char *graphName, WUGraphState state) const
+    {
+        setNodeState(graphName, 0, state);
+    }
+    void setNodeState(const char *graphName, WUGraphIDType nodeId, WUGraphState state) const
+    {
+        CassandraStatement statement(sessionCache->prepareStatement("INSERT INTO wuGraphState (partition, wuid, graphID, subgraphID, state) values (?,?,?,?,?);"));
```

Comment:
If this is called once for running, and again for complete, doesn't that create 2 entries?  Is that ok?